File: apps/dsa-web/src/api/scanner.ts
import apiClient from './index';
import { toCamelCase } from './utils';

// ============ Types ============

export interface ScannerRunRequest {
  market?: string;
  biasThreshold?: number;
  gainMin?: number;
  gainMax?: number;
  volumeRatioMin?: number;
  scoreThreshold?: number;
  topQuant?: number;
  topFinal?: number;
}

export interface ScannerRunResponse {
  taskId: string;
}

export interface ScannerStatusResponse {
  taskId: string;
  status: 'pending' | 'scanning' | 'llm' | 'done' | 'error';
  progress: number;
  total: number;
  screened: number;
  message: string;
  error?: string;
}

export interface ScannerCandidateItem {
  code: string;
  name: string;
  market: string;
  quantScore: number;
  maScore: number;
  biasScore: number;
  volumeScore: number;
  gainScore: number;
  currentPrice?: number;
  ma5?: number;
  ma10?: number;
  ma20?: number;
  biasMa5?: number;
  volumeRatio?: number;
  gain20d?: number;
  peRatio?: number;
  pbRatio?: number;
  llmRank?: number;
  llmReason?: string;
  llmSelected: boolean;
  confirmed: boolean;
}

export interface ScannerCandidatesResponse {
  taskId?: string;
  candidates: ScannerCandidateItem[];
  total: number;
}

export interface ScannerConfirmResponse {
  confirmed: number;
  added: string[];
  stockList: string[];
}

// ============ API ============

export const scannerApi = {
  run: async (params: ScannerRunRequest = {}): Promise<ScannerRunResponse> => {
    const requestData: Record<string, unknown> = {};
    if (params.market) requestData.market = params.market;
    if (params.biasThreshold != null) requestData.bias_threshold = params.biasThreshold;
    if (params.gainMin != null) requestData.gain_min = params.gainMin;
    if (params.gainMax != null) requestData.gain_max = params.gainMax;
    if (params.volumeRatioMin != null) requestData.volume_ratio_min = params.volumeRatioMin;
    if (params.scoreThreshold != null) requestData.score_threshold = params.scoreThreshold;
    if (params.topQuant != null) requestData.top_quant = params.topQuant;
    if (params.topFinal != null) requestData.top_final = params.topFinal;

    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/scanner/run',
      requestData,
    );
    return toCamelCase<ScannerRunResponse>(response.data);
  },

  getStatus: async (taskId: string): Promise<ScannerStatusResponse> => {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/scanner/status/${taskId}`,
    );
    return toCamelCase<ScannerStatusResponse>(response.data);
  },

  getCandidates: async (params: {
    taskId?: string;
    llmOnly?: boolean;
  } = {}): Promise<ScannerCandidatesResponse> => {
    const queryParams: Record<string, string | boolean> = {};
    if (params.taskId) queryParams.task_id = params.taskId;
    if (params.llmOnly != null) queryParams.llm_only = params.llmOnly;

    const response = await apiClient.get<Record<string, unknown>>(
      '/api/v1/scanner/candidates',
      { params: queryParams },
    );
    return toCamelCase<ScannerCandidatesResponse>(response.data);
  },

  confirm: async (taskId: string, codes: string[]): Promise<ScannerConfirmResponse> => {
    const response = await apiClient.post<Record<string, unknown>>(
      '/api/v1/scanner/confirm',
      { task_id: taskId, codes },
    );
    return toCamelCase<ScannerConfirmResponse>(response.data);
  },
};

File: api/v1/schemas/scanner.py
# -*- coding: utf-8 -*-
"""Scanner API schemas."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class ScannerRunRequest(BaseModel):
    market: str = Field("us", description="market to scan: us / cn")
    bias_threshold: Optional[float] = Field(None, ge=1.0, le=20.0, description="乖离率上限(%)")
    gain_min: Optional[float] = Field(None, ge=0.0, description="近20日涨幅下限(%)")
    gain_max: Optional[float] = Field(None, ge=5.0, description="近20日涨幅上限(%)")
    volume_ratio_min: Optional[float] = Field(None, ge=0.5, description="量比下限")
    score_threshold: Optional[int] = Field(None, ge=10, le=100, description="进LLM的最低得分")
    top_quant: Optional[int] = Field(None, ge=5, le=100, description="量化初筛后进LLM的数量")
    top_final: Optional[int] = Field(None, ge=1, le=50, description="LLM精选最终输出数量")


class ScannerRunResponse(BaseModel):
    task_id: str = Field(..., description="异步任务ID")


class ScannerStatusResponse(BaseModel):
    task_id: str
    status: str = Field(..., description="pending/scanning/llm/done/error")
    progress: int = Field(0, description="已处理数量")
    total: int = Field(0, description="总数量")
    screened: int = Field(0, description="通过初筛数量")
    message: str = Field("", description="当前状态描述")
    error: Optional[str] = Field(None, description="错误信息")


class ScannerCandidateItem(BaseModel):
    code: str
    name: str
    market: str = "us"
    quant_score: float = Field(0, description="量化总分(0-100)")
    ma_score: float = 0
    bias_score: float = 0
    volume_score: float = 0
    gain_score: float = 0
    current_price: Optional[float] = None
    ma5: Optional[float] = None
    ma10: Optional[float] = None
    ma20: Optional[float] = None
    bias_ma5: Optional[float] = None
    volume_ratio: Optional[float] = None
    gain_20d: Optional[float] = None
    pe_ratio: Optional[float] = None
    pb_ratio: Optional[float] = None
    llm_rank: Optional[int] = None
    llm_reason: Optional[str] = None
    llm_selected: bool = False
    confirmed: bool = False


class ScannerCandidatesResponse(BaseModel):
    task_id: Optional[str] = None
    candidates: List[ScannerCandidateItem] = Field(default_factory=list)
    total: int = 0


class ScannerConfirmRequest(BaseModel):
    task_id: str = Field(..., description="任务ID")
    codes: List[str] = Field(..., min_length=1, description="确认加入自选的股票代码列表")


class ScannerConfirmResponse(BaseModel):
    confirmed: int = Field(0, description="确认数量")
    added: List[str] = Field(default_factory=list, description="实际新增的股票代码")
    stock_list: List[str] = Field(default_factory=list, description="更新后的完整STOCK_LIST")
